public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx

```diff
@@ -0,0 +1,493 @@
+import { LokiLabel } from './LokiLabel';
+
+// Hard limit on labels to render
+const MAX_LABEL_COUNT = 100;
+const MAX_VALUE_COUNT = 10000;
+const MAX_AUTO_SELECT = 4;
+const EMPTY_SELECTOR = '{}';
+export const LAST_USED_LABELS_KEY = 'grafana.datasources.loki.browser.labels';
+
+type onChange = (selector: string) => void;
+
+export interface BrowserProps {
+  languageProvider: LokiLanguageProvider;
+  onChange: onChange;
+  theme: GrafanaTheme;
+  autoSelect?: number;
+  hide?: () => void;
+}
+
+interface BrowserState {
+  labels: SelectableLabel[];
+  searchTerm: string;
+  status: string;
+  error: string;
+  validationStatus: string;
+}
+
+interface FacettableValue {
+  name: string;
+  // facetted?: boolean; // True if value is possible given the selector
+  selected?: boolean;
+}
+
+export interface SelectableLabel {
+  name: string;
+  selected?: boolean;
+  loading?: boolean;
+  values?: FacettableValue[];
+  hidden?: boolean;
+  facets?: number;
+}
+
+export function buildSelector(labels: SelectableLabel[]): string {
+  const selectedLabels = [];
+  for (const label of labels) {
+    if (label.selected && label.values && label.values.length > 0) {
+      const selectedValues = label.values.filter((value) => value.selected).map((value) => value.name);
+      if (selectedValues.length > 1) {
+        selectedLabels.push(`${label.name}=~"${selectedValues.join('|')}"`);
+      } else if (selectedValues.length === 1) {
+        selectedLabels.push(`${label.name}="${selectedValues[0]}"`);
+      }
+    }
+  }
+  return ['{', selectedLabels.join(','), '}'].join('');
+}
+
+export function facetLabels(
+  labels: SelectableLabel[],
+  possibleLabels: Record<string, string[]>,
+  lastFacetted?: string
+): SelectableLabel[] {
+  return labels.map((label) => {
+    const possibleValues = possibleLabels[label.name];
+    if (possibleValues) {
+      let existingValues: FacettableValue[];
+      if (label.name === lastFacetted && label.values) {
+        // Facetting this label, show all values
+        existingValues = label.values;
+      } else {
+        // Keep selection in other facets
+        const selectedValues: Set<string> = new Set(
+          label.values?.filter((value) => value.selected).map((value) => value.name) || []
+        );
+        // Values for this label have not been requested yet, let's use the facetted ones as the initial values
+        existingValues = possibleValues.map((value) => ({ name: value, selected: selectedValues.has(value) }));
+      }
+      return { ...label, loading: false, values: existingValues, facets: existingValues.length };
+    }
+
+    // Label is facetted out, hide all values
+    return { ...label, loading: false, hidden: !possibleValues, values: undefined, facets: 0 };
+  });
+}
+
+const getStyles = stylesFactory((theme: GrafanaTheme) => ({
+  wrapper: css`
+    background-color: ${theme.colors.bg2};
+    padding: ${theme.spacing.md};
+    max-width: 100rem;
```

Comment:
Is there a reason why we don't do `width: 100%` as we are in query editor - we can use the full provided witdth. Here (if not that much labels) the width is too small and "ready" text is in the way of the button. 
![image](https://user-images.githubusercontent.com/30407135/109535603-0cf93b80-7abd-11eb-9969-8655d1c310dd.png)

Reply:
Also I am not sure if `max-width: 100rem;` does anything because when I am adding labels to search in, the label browser window is growing and growing and pushing query editor outside the view, so I have to scroll.
![image](https://user-images.githubusercontent.com/30407135/109537386-1daab100-7abf-11eb-9c00-f53dcdd63fd8.png)